flink-runtime/src/test/java/org/apache/flink/runtime/io/network/netty/NettyLeakDetectionResource.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.io.network.netty;
+
+import org.apache.flink.shaded.netty4.io.netty.util.ResourceLeakDetector;
+import org.apache.flink.shaded.netty4.io.netty.util.ResourceLeakDetectorFactory;
+
+import org.junit.rules.ExternalResource;
+
+import static org.junit.Assert.fail;
+
+/**
+ * JUnit resource to fail with an assertion when Netty detects a resource leak (only with
+ * <tt>ERROR</tt> logging enabled for
+ * <tt>org.apache.flink.shaded.netty4.io.netty.util.ResourceLeakDetector</tt>).
+ *
+ * <p>This should be used in a class rule:
+ * <pre>{@code
+ * @literal @ClassRule
+ *  public static final NettyLeakDetectionResource LEAK_DETECTION = new NettyLeakDetectionResource();
+ * }</pre>
+ */
+public class NettyLeakDetectionResource extends ExternalResource {
+	private static ResourceLeakDetectorFactory previousLeakDetector;
+	private static ResourceLeakDetector.Level previousLeakDetectorLevel;
+
+	@Override
+	protected void before() {
+		previousLeakDetector = ResourceLeakDetectorFactory.instance();
+		previousLeakDetectorLevel = ResourceLeakDetector.getLevel();
+
+		ResourceLeakDetector.setLevel(ResourceLeakDetector.Level.PARANOID);
+		ResourceLeakDetectorFactory.setResourceLeakDetectorFactory(new FailingResourceLeakDetectorFactory());
```

Comment:
so this isn't something we necessarily have to deal with now, but if multiple tests that use this resource run in parallel we might not reset to the correct factory/level.

Effectively what we need is some kinda ref-counting so that only the first resource modifies the level and factory, and only the last resource reset them. :/

Reply:
true - done